android/src/main/java/io/github/elyx0/reactnativedocumentpicker/DocumentPickerModule.java

```diff
@@ -175,6 +184,75 @@ public void onShowActivityResult(int resultCode, Intent data, Promise promise) {
 		}
 	}
 
+	@ReactMethod
+	public void pickDirectory(Promise promise) {
+		Activity currentActivity = getCurrentActivity();
+
+		if (currentActivity == null) {
+			promise.reject(E_ACTIVITY_DOES_NOT_EXIST, "Current activity does not exist");
+			return;
+		}
+		this.promise = promise;
+		try {
+			Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
+			currentActivity.startActivityForResult(intent, PICK_DIR_REQUEST_CODE, null);
+		} catch (Exception e) {
+			sendError(E_FAILED_TO_SHOW_PICKER, "Failed to create directory picker", e);
+		}
+	}
+
+	private void onPickDirectoryResult(int resultCode, Intent data, Promise promise) {
+		if (resultCode == Activity.RESULT_CANCELED) {
+			sendError(E_DOCUMENT_PICKER_CANCELED, "User canceled directory picker");
+			return;
+		} else if (resultCode != Activity.RESULT_OK) {
+			sendError(E_UNKNOWN_ACTIVITY_RESULT, "Unknown activity result: " + resultCode);
+			return;
+		}
+
+		if (data == null || data.getData() == null) {
+			sendError(E_INVALID_DATA_RETURNED, "Invalid data returned by intent");
+			return;
+		}
+		Uri uri = data.getData();
+
+		WritableMap map = Arguments.createMap();
+		map.putString(FIELD_URI, uri.toString());
+
+		String path = uriToPath(uri);
+		if (path != null) {
+			map.putString(FIELD_PATH, path);
+		}
+		promise.resolve(map);
+	}
+
+	private String uriToPath(Uri uri) {
+		if (ContentResolver.SCHEME_FILE.equals(uri.getScheme())) {
+			File file = new File(uri.getPath());
+			return file.getName();
+		} else if (ContentResolver.SCHEME_CONTENT.equals(uri.getScheme())) {
+			String name = null;
+
+			// URI examples
+			// internal: content://com.android.externalstorage.documents/tree/primary%3Atest
+			// sd card:  content://com.android.externalstorage.documents/tree/1DEA-0313%3Atest
+			List<String> pathSegments = uri.getPathSegments();
+			if (pathSegments.get(0).equalsIgnoreCase("tree")) {
```

Comment:
did you write this code by yourself, or is it copied from somewhere (where?)

I'm worried this can vary a lot among android versions and I'm not sure this would work reliably... 🤔

Reply:
I have wrote it myself. I could only test it with emulator and my phone (Honor 9) which does not have external sd card.

Reply:
Hello and thanks for the reply :) I think I need more context to understand.

What is the reason for returning the `path` from this call? Why not return only the `uri`?

If the reason is that you want to write a file to that path later on, then I believe the right approach to do this is outlined here https://stackoverflow.com/a/61120265/2070942

The approach taken here seems like a heuristic that works for some cases, but may not work in others so I'm hesitant to merging this. We should be using android-system-provided tools to work with uris.

If the only thing returned from the call were the uri, then I'd be happy with this.

Reply:
> What is the reason for returning the path from this call? Why not return only the uri?

This is so that you can then use react-native-fs to access the directory. It may stop working in Android 12 but I suppose it's better than nothing.

> If the reason is that you want to write a file to that path later on, then I believe the right approach to do this is outlined here https://stackoverflow.com/a/61120265/2070942 and should not be handled in this library.

Sure, but then you won't be using react-native any more and would be just writing a native app.

> If the only thing returned from the call were the uri, then I'd be happy with this.

As far as I'm aware you can't do anything with that URI from RN, without going native.

Reply:
> The approach taken here seems like a heuristic that works for some cases, but may not work in others so I'm hesitant to merging this. We should be using android-system-provided tools to work with uris.

Ok, fair enough.